### PR TITLE
[FEAT] Add `.str.split()` API for splitting string columns.

### LINKED ISSUE #1388: Add .str.split() expression
**Is your feature request related to a problem? Please describe.**

Add a `.str.split(delim)` expression which splits a String column into `List[string]` based on a delimiter

### PR DESCRIPTION
**Pull request by clarkzinzow**
This PR adds an `Expression.str.split()` API for splitting strings in string columns on a pattern.

Closes #1388

### CODE REVIEW

File: src/daft-core/src/array/ops/full.rs [hunk 1]
```diff
@@ -129,7 +129,8 @@ impl FullNull for ListArray {
                 Self::new(
                     Field::new(name, dtype.clone()),
                     empty_flat_child,
-                    OffsetsBuffer::try_from(repeat(0).take(length).collect::<Vec<_>>()).unwrap(),
+                    OffsetsBuffer::try_from(repeat(0).take(length + 1).collect::<Vec<_>>())
```
Comment (clarkzinzow) [inline src/daft-core/src/array/ops/full.rs:132]: @jaychia FYI I made this fix as a driveby, where the offset buffer should be one element longer than the data array and validity mask.

**Review by jaychia — APPROVED** [review]
Mostly lgtm, added a few comments. Nice tests!
